File: LeetCode/93-restore-ip-addresses/Solution.java
import java.util.ArrayList;
import java.util.List;

/**
 * This class provides a method to restore valid IP addresses from a given
 * string.
 */
public class Solution {

  /**
   * Restores valid IP addresses from the given string.
   *
   * @param s the input string containing only digits
   * @return a list of all possible valid IP address combinations
   */
  public static List<String> restoreIpAddresses(String s) {
    List<String> ans = new ArrayList<>();
    backtrack(s, 0, new ArrayList<>(), ans);
    return ans;
  }

  /**
   * Utility method to backtrack and find all valid IP address combinations.
   *
   * @param s          the input string containing only digits
   * @param startIndex the current index to start processing
   * @param dots       a list of integers representing the positions of the dots
   *                   added so far
   * @param ans        a list of strings to store the valid IP address
   *                   combinations
   */
  private static void backtrack(String s, int startIndex, List<Integer> dots, List<String> ans) {
    int remainingLength = s.length() - startIndex;
    int remainingNumberOfIntegers = 4 - dots.size();

    // Check if the remaining length is too long or too short to form a valid IP
    // address
    if (remainingLength > remainingNumberOfIntegers * 3 || remainingLength < remainingNumberOfIntegers) {
      return;
    }

    // If we are adding the last segment
    if (remainingNumberOfIntegers == 1) {
      if (validSegment(s, startIndex, remainingLength)) {
        StringBuilder ipAddress = new StringBuilder();
        int last = 0;
        // Build the IP address with dots
        for (int dot : dots) {
          ipAddress.append(s, last, last + dot).append('.');
          last += dot;
        }
        ipAddress.append(s, startIndex, s.length());
        ans.add(ipAddress.toString());
      }
      return;
    }

    // Try to place a dot after every 1 to 3 digits
    for (int curPos = 1; curPos <= Math.min(3, remainingLength); curPos++) {
      if (validSegment(s, startIndex, curPos)) {
        dots.add(curPos);
        backtrack(s, startIndex + curPos, dots, ans);
        dots.remove(dots.size() - 1);
      }
    }
  }

  /**
   * Checks if the segment of the string is a valid IP address segment.
   *
   * @param s      the input string
   * @param start  the starting index of the segment
   * @param length the length of the segment
   * @return true if the segment is valid, false otherwise
   */
  private static boolean validSegment(String s, int start, int length) {
    if (length == 1) {
      return true;
    }
    if (s.charAt(start) == '0') {
      return false;
    }
    if (length == 2) {
      return 10 <= Integer.parseInt(s.substring(start, start + 2))
          && Integer.parseInt(s.substring(start, start + 2)) <= 99;
    }
    if (length == 3) {
      return 100 <= Integer.parseInt(s.substring(start, start + 3))
          && Integer.parseInt(s.substring(start, start + 3)) <= 255;
    }
    return false;
  }
}

File: LeetCode/93-restore-ip-addresses/solution.ts
/**
 * Restores IP addresses from a given string using backtracking.
 * @param {string} s - The input string containing digits.
 * @returns {string[]} - An array of valid IP addresses.
 */
function restoreIpAddresses(s: string): string[] {
  const result: string[] = [];
  backtrack(s, 0, [], result);
  return result;
};

/**
 * Helper function to perform backtracking and restore IP addresses.
 * @param {string} s - The input string.
 * @param {number} startIndex - The current start index in the string.
 * @param {number[]} dots - List of segment lengths.
 * @param {string[]} result - List to store the resulting IP addresses.
 */
function backtrack(
  s: string,
  startIndex: number,
  dots: number[],
  result: string[]
): void {
  const remainingLength = s.length - startIndex;
  const remainingNumberOfIntegers = 4 - dots.length;

  // Check if the remaining characters can form valid segments.
  if (
    remainingLength > remainingNumberOfIntegers * 3 ||
    remainingLength < remainingNumberOfIntegers
  ) {
    return;
  }

  // If we have 1 segment left to place.
  if (remainingNumberOfIntegers === 1) {
    if (isValidSegment(s, startIndex, remainingLength)) {
      let ipAddress = "";
      let last = 0;
      // Construct the IP address from the segments.
      for (const dot of dots) {
        ipAddress += s.slice(last, last + dot) + ".";
        last += dot;
      }
      ipAddress += s.slice(startIndex);
      result.push(ipAddress);
    }
    return;
  }

  // Try placing a dot at every possible position.
  for (let curPos = 1; curPos <= Math.min(3, remainingLength); curPos++) {
    if (isValidSegment(s, startIndex, curPos)) {
      dots.push(curPos);
      backtrack(s, startIndex + curPos, dots, result);
      dots.pop();
    }
  }
}

/**
 * Validates whether the segment of the string is a valid IP address segment.
 * @param {string} s - The input string.
 * @param {number} start - The start index of the segment.
 * @param {number} length - The length of the segment.
 * @returns {boolean} - True if the segment is valid, false otherwise.
 */
function isValidSegment(s: string, start: number, length: number): boolean {
  if (length === 1) {
    return true;
  }
  if (s.charAt(start) === "0") {
    return false;
  }
  if (length === 2) {
    const segment = parseInt(s.slice(start, start + 2), 10);
    return segment >= 10 && segment <= 99;
  }
  if (length === 3) {
    const segment = parseInt(s.slice(start, start + 3), 10);
    return segment >= 100 && segment <= 255;
  }
  return false;
}
